Public/Src/Cache/ContentStore/Distributed/Sessions/DistributedContentSession.cs

```diff
@@ -165,5 +180,51 @@ private async Task<PutResult> RegisterPutAsync(Context context, CancellationToke
 
             return putResult;
         }
+
+        private async Task<BoolResult> RequestProactiveCopyIfNeededAsync(OperationContext context, ContentHash hash)
+        {
+            var hashArray = new[] { hash };
+
+            // First check in local location store, then global if failed.
+            var getLocationsResult = await ContentLocationStore.GetBulkAsync(context, hashArray, context.Token, UrgencyHint.Nominal, GetBulkOrigin.Local);
+            if (!getLocationsResult.Succeeded || getLocationsResult.ContentHashesInfo[0].Locations.Count == 0)
+            {
+                getLocationsResult = await ContentLocationStore.GetBulkAsync(context, hashArray, context.Token, UrgencyHint.Nominal, GetBulkOrigin.Global);
+
+                if (getLocationsResult.Succeeded)
+                {
+                    _counters[Counters.GetLocationsSatisfiedFromRemote].Increment();
+                }
+            }
+            else
+            {
+                _counters[Counters.GetLocationsSatisfiedFromLocal].Increment();
+            }
+
+            if (getLocationsResult.Succeeded)
+            {
+                // Only request copy when this is the only machine where the content is located.
+                if (getLocationsResult.ContentHashesInfo[0].Locations.Count <= 1)
+                {
+                    var location = ContentLocationStore.GetRandomMachineLocation(except: LocalCacheRootMachineLocation);

```

Comment:
We need another counter to track the following more explicitely (in the trace message for sure and, maybe in the counters):

The result of this operation is:
we either:
- requested a copy because there is only N copies available (this threshold should be configurable)
- did not requested a copy because we have enough copies
- did not requested a copy because we failed to find a machine in the stamp.

I would suggest to restructure this in the fllowing way (-ish):

var getLocations = GetLocationsFromMultipleLevels(context, hash);

if (!getLocations) {
  // We failed to get locations, but we still (potentially) may try to do a proactive copy if we want!
  return new BoolResult($"Failed to proactively copy the hash '{hash}'. {getLocations}");
}

if (getLocations.Origin == Local) {
  incrementCounter for locals;
}
else {
  incrmenent counter for global;
}

if (getLocations.Locations.Count < ConfiguredThreashold) {
  // assuming GetRandomMachineLocations returns BoolResult.
   var targetLocation = GetRandomMachineLocation();
  if (!targetLocation) {
    return new BoolResult($"Failed to proactively copy the hash '{hash}'. {targetLocation});
  }

  return await DistributedCopier.RequestCopyFileAsync();
}